plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,402 @@
+  where
+    addToMap ::
+      Term name uni fun (Path, ann) ->
+      HashMap (Term name uni fun ()) (HashMap Path (Term name uni fun (Path, ann), Int)) ->
+      HashMap (Term name uni fun ()) (HashMap Path (Term name uni fun (Path, ann), Int))
+    addToMap t0
+      -- We don't consider work-free terms for CSE, because doing so may or may not
+      -- have a size benefit, but certainly doesn't have any cost benefit (the cost
+      -- will in fact be slightly higher due to the additional application).
+      --
+      -- `isWorkFree` currently doesn't check whether a builtin application is saturated,
+      -- or whether an term is the (possibly repeated) forcing of a builtin (which should
+      -- be workfree), so we check it separately.
+      | isWorkFree t0 || not (isBuiltinSaturated t0) || isForcingBuiltin t0 = id
+      | otherwise =
+          Map.alter
+            ( \case
+                Nothing    -> Just $ Map.singleton path (t0, 1)
+                Just paths -> Just $ combinePaths t0 path paths
+            )
+            t
+      where
+        t = void t0
+        path = fst (termAnn t0)
+
+    isBuiltinSaturated = \case
+      t@Apply{}
+        | (Builtin _ fun, args) <- splitApplication t ->
+            length args >= Map.findWithDefault 0 fun arityInfo
+      _ -> True
+
+    isForcingBuiltin = \case
+      Builtin{} -> True
+      Force _ t -> isForcingBuiltin t
+      _         -> False
+
+-- | Combine a new path with a number of existing (path, count) pairs.
+combinePaths ::
+  forall name uni fun ann.
+  Term name uni fun (Path, ann) ->
+  Path ->
+  HashMap Path (Term name uni fun (Path, ann), Int) ->
+  HashMap Path (Term name uni fun (Path, ann), Int)
+combinePaths t path = Map.fromListWith (\(x, y) (_, y') -> (x, y + y')) . go . Map.toList
+  where
+    go ::
+      [(Path, (Term name uni fun (Path, ann), Int))] ->
+      [(Path, (Term name uni fun (Path, ann), Int))]
+    -- The new path is not a descendent-or-self of any existing path.
+    go [] = [(path, (t, 1))]
+    go ((path', (t', cnt)) : paths)
+      -- The new path is an ancestor-or-self of an existing path.
+      -- Take over all counts of the existing path, remove the existing path,
+      -- and continue.
+      | path `isAncestorOrSelf` path' = (path, (t, cnt)) : go paths
+      -- The new path is a descendent-or-self of an existing path.
+      -- Increment the count for the existing path. There can only be one such
+      -- existing path, so we don't need to recurse here.
+      | path' `isAncestorOrSelf` path = (path', (t', cnt + 1)) : paths
+      | otherwise = (path', (t', cnt)) : go paths
+
+mkCseTerm ::
+  forall uni fun ann m.
+  (MonadQuote m, Eq (Term Name uni fun ())) =>
+  [Term Name uni fun (Path, ann)] ->
+  -- | The original annotated term
+  Term Name uni fun (Path, ann) ->
+  m (Term Name uni fun ann)
+mkCseTerm ts t = do
+  cs <- traverse mkCseCandidate ts
+  pure . fmap snd $ foldl' (flip applyCse) t cs
```

Comment:
Why not handle the whole list of candidates in `applyCse` instead of doing it one-by-one? Doesn't seem to be much more complicated and will save you some runtime. Maybe it doesn't matter.

Reply:
> Why not handle the whole list of candidates in applyCse instead of doing it one-by-one?

It can be bit tricky to implement (if not a lot trickier). Different CSE candidates need to be inserted into different places in the original term. And different CSE candidates may contain each other. It is not immediately obvious to me what an elegant way of doing it might look like.

Reply:
Fair enough.